// gpsoauth-js/dist/gpsoauth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GPSOAuth = void 0;
// Android key from gpsoauth version 7.3.29
const B64_KEY_7_3_29 = "AAAAgMom/1a/v0lblO2Ubrt60J2gcuXSljGFQXgcyZWveWLEwo6prwgi3" +
    "iJIZdodyhKZQrNWp5nKJ3srRXcUW+F1BD3baEVGcmEgqaLZUNBjm057pK" +
    "RI16kB0YppeGx5qIQ5QjKzsR8ETQbKLNWgRY0QRNVz34kMJR3P/LgHax/" +
    "6rmf5AAAAAwEAAQ==";
const AUTH_URL = "https://android.clients.google.com/auth";
/**
 * Modern JavaScript/TypeScript implementation of Google Play Services OAuth (gpsoauth)
 * Compatible with React Native and Node.js environments
 */
class GPSOAuth {
    /**
     * Exchange web oauth token for master token
     */
    static async exchangeToken(email, token, config = {}) {
        const { service = "ac2dm", androidId = GPSOAuth.generateAndroidId(), deviceCountry = "us", operatorCountry = "us", lang = "en", sdkVersion = 17 } = config;
        try {
            console.log('🔄 Starting token exchange...');
            const formData = new URLSearchParams();
            formData.append('accountType', 'HOSTED_OR_GOOGLE');
            formData.append('Email', email);
            formData.append('has_permission', '1');
            formData.append('add_account', '1');
            formData.append('ACCESS_TOKEN', '1');
            formData.append('Token', token);
            formData.append('service', service);
            formData.append('source', 'android');
            formData.append('androidId', androidId);
            formData.append('device_country', deviceCountry);
            formData.append('operatorCountry', operatorCountry);
            formData.append('lang', lang);
            formData.append('sdk_version', sdkVersion.toString());
            formData.append('client_sig', '38918a453d07199354f8b19af05ec6562ced5788');
            formData.append('callerSig', '38918a453d07199354f8b19af05ec6562ced5788');
            const response = await GPSOAuth.makeAuthRequest(formData);
            if (response.error) {
                return response;
            }
            return {
                masterToken: response.token,
                email: response.email,
                auth: response.auth,
            };
        }
        catch (error) {
            console.error('💥 Token exchange exception:', error);
            return { error: `Token exchange failed: ${error instanceof Error ? error.message : 'Unknown error'}` };
        }
    }
    /**
     * Perform OAuth to a specific Google service using master token
     */
    static async performOAuth(email, masterToken, config = {}) {
        const { service = 'sj', app = 'com.google.android.music', clientSig = '...', androidId = GPSOAuth.generateAndroidId(), deviceCountry = "us", operatorCountry = "us", lang = "en", sdkVersion = 17, googlePlayServicesVersion = 240913000 } = config;
        try {
            console.log('🔓 Starting OAuth authentication...');
            const formData = new URLSearchParams();
            formData.append('accountType', 'HOSTED_OR_GOOGLE');
            formData.append('Email', email);
            formData.append('has_permission', '1');
            formData.append('EncryptedPasswd', masterToken);
            formData.append('service', service);
            formData.append('source', 'android');
            formData.append('androidId', androidId);
            formData.append('app', app);
            formData.append('client_sig', clientSig);
            formData.append('device_country', deviceCountry);
            formData.append('operatorCountry', operatorCountry);
            formData.append('lang', lang);
            formData.append('sdk_version', sdkVersion.toString());
            formData.append('google_play_services_version', googlePlayServicesVersion.toString());
            // Log form data for debugging
            console.log('📋 Complete OAuth form data:');
            for (const [key, value] of formData.entries()) {
                if (key === 'EncryptedPasswd') {
                    console.log(`  ${key}: ${value.substring(0, 20)}...`);
                }
                else {
                    console.log(`  ${key}: ${value}`);
                }
            }
            const response = await GPSOAuth.makeAuthRequest(formData);
            if (response.error) {
                return response;
            }
            return {
                masterToken: response.auth,
                email: response.email,
                auth: response.auth,
                lsid: response.lsid,
                sid: response.sid,
            };
        }
        catch (error) {
            console.error('💥 OAuth exception:', error);
            return { error: `OAuth failed: ${error instanceof Error ? error.message : 'Unknown error'}` };
        }
    }
    /**
     * Make HTTP request to Google's auth endpoint
     */
    static async makeAuthRequest(formData) {
        const requestHeaders = {
            'User-Agent': 'GoogleAuth/1.4 (gzip)',
            'Content-Type': 'application/x-www-form-urlencoded',
            'Accept-Encoding': 'identity',
            'Accept': '*/*',
            'Connection': 'Keep-Alive',
        };
        console.log('🌐 Making request to:', AUTH_URL);
        const response = await fetch(AUTH_URL, {
            method: 'POST',
            headers: requestHeaders,
            body: formData.toString(),
        });
        const responseText = await response.text();
        console.log('📥 Response status:', response.status);
        if (!response.ok) {
            console.error('❌ Auth response error:', response.status);
            return { error: `Authentication failed: ${response.status} ${response.statusText}` };
        }
        return GPSOAuth.parseAuthResponse(responseText);
    }
    /**
     * Parse Google's authentication response
     */
    static parseAuthResponse(responseText) {
        const result = {};
        const lines = responseText.split('\n');
        for (const line of lines) {
            if (line.trim() === '')
                continue;
            if (line.includes('=')) {
                const [key, value] = line.split('=', 2);
                const trimmedKey = key.trim().toLowerCase();
                const trimmedValue = value.trim();
                result[trimmedKey] = trimmedValue;
            }
        }
        if (result.error) {
            return { error: `Google auth error: ${result.error}` };
        }
        if (result.errormsg) {
            return { error: `Google auth error: ${result.errormsg}` };
        }
        return result;
    }
    /**
     * Generate a mock Android device ID
     */
    static generateAndroidId() {
        const chars = '0123456789abcdef';
        let result = '';
        for (let i = 0; i < 16; i++) {
            result += chars.charAt(Math.floor(Math.random() * chars.length));
        }
        return result;
    }
}
exports.GPSOAuth = GPSOAuth;
